src/ILVerify/src/ILImporter.StackValue.cs

```diff
@@ -247,6 +247,232 @@ public static TypeDesc GetIntermediateType(TypeDesc type)
             }
         }
 
+        /// <summary>
+        /// Merges two stack values to a common stack value as defined in the ECMA-335 
+        /// standard III.1.8.1.3 (Merging stack states).
+        /// </summary>
+        /// <param name="valueA">The value to be merged with <paramref name="valueB"/>.</param>
+        /// <param name="valueB">The value to be merged with <paramref name="valueA"/>.</param>
+        /// <param name="merged">The resulting type of merging <paramref name="valueA"/> and <paramref name="valueB"/>.</param>
+        /// <returns>True if merge operation was successful, false if the merge operation failed.</returns>
+        public static bool TryMergeStackValues(StackValue valueA, StackValue valueB, out StackValue merged)
+        {
+            merged = valueA;
+
+            if (valueB.IsReadOnly)
+                merged.SetIsReadOnly();
+
+            // Same type
+            if (valueA.Kind == valueB.Kind && valueA.Type == valueB.Type)
+                return true;
+
+            // One of the two types is a (unboxed) generic parameter (only valid if a == b)
+            if ((valueA.Type != null && valueA.Type.IsGenericParameter && valueA.Kind != StackValueKind.ObjRef) ||
+                (valueB.Type != null && valueB.Type.IsGenericParameter && valueB.Kind != StackValueKind.ObjRef))
+                return false;
+
+            if (valueA.IsNullReference)
+            {
+                //Null can be any reference type
+                if (valueB.Kind == StackValueKind.ObjRef)
+                {
+                    merged = valueB;
+                    return true;
+                }
+            }
+            else if (valueA.Kind == StackValueKind.ObjRef)
+            {
+                if (valueB.Kind != StackValueKind.ObjRef)
+                    return false;
+
+                // Null can be any reference type
+                if (valueB.IsNullReference)
+                    return true;
+
+                // Merging classes always succeeds since System.Object always works
+                merged = StackValue.CreateFromType(MergeClasses(valueA.Type, valueB.Type)); 
+                return true;
+            }
+
+            return false;
+        }
+
+        // Used to merge stack states.
+        static TypeDesc MergeClasses(TypeDesc classA, TypeDesc classB)
+        {
+            if (classA == classB)
+                return classA;
+
+            // Array case
+            if (classA.IsArray)
+            {
+                if (classB.IsArray)
+                    return MergeArrayTypes((ArrayType)classA, (ArrayType)classB);
+                classA = classA.Context.GetWellKnownType(WellKnownType.Array);
```

Comment:
I do not think the actual array type should be replace by the more general array type here. It may cause false negatives for the IsGenericParameter check below, e.g. if merging `int[]` and generic parameter with `int[]` constrain.

Reply:
You are right, this replacement is unnecessary.
I would like to add a test case for this kind of scenario, however I am not aware of any way to constrain a generic with `int[]`.

Reply:
C# disallows it, but it should be possible to write it in IL directly: `.class private auto ansi beforefieldinit G`1<(string[]) T>`

Reply:
Ah, I see, thanks.
I implemented the test case and removed the replacement with `System.Array`.